tools/rebalance-corenrn-data.py

```diff
@@ -0,0 +1,231 @@
+
+    for dat_entry, size in entry_sizes:
+        try:
+            distribute_dat_to_bucket(dat_entry, size, buckets, bucket_heap)
+        except Exception as e:
+            raise RuntimeError(f"Error processing dat entry {dat_entry}") from e
+
+    if show_stats:
+        logging.info("Top 10 machines accumulated sizes")
+        for size, mach_i in heapq.nlargest(10, bucket_heap):
+            print(f"  Machine {mach_i}: {size/(1024*1024):.1f} MiB")
+
+        mach_sizes = [bucket[0] for bucket in bucket_heap]
+        show_histogram(mach_sizes)
+
+    return buckets, metadata
+
+
+def write_dat_file(buckets, infos: dict, ranks_per_machine, output_file="rebalanced-files.dat"):
+    """
+    Output the result after processing all directories
+    """
+    DEFAULT_LINE = f"{CORENRN_SKIP_MARK}\n" * ranks_per_machine
+    logging.info("Writing out data from %d buckets to file: %s", len(buckets), output_file)
+
+    # CoreNeuron does RoundRobin - we need to transpose the entries
+    # When a sequence finishes use "-1" (to keep in sync)
+
+    def batch(iterable, first=0):
+        last = first + ranks_per_machine
+        group = iterable[first:last]
+        while group:
+            if len(group) < ranks_per_machine:
+                yield group + [CORENRN_SKIP_MARK] * (ranks_per_machine - len(group))
+                break
+            yield group
+            first, last = last, last + 40
+            group = iterable[first:last]
+
+    with open(output_file, "w") as out:
+        print(infos["version"], file=out)
+        print(infos["n_files"], file=out)
+
+        for buckets in itertools.zip_longest(*[batch(m) for m in buckets]):
+            for entries in buckets:
+                if entries is None:
+                    out.write(DEFAULT_LINE)
+                else:
+                    for entry in entries:
+                        out.write(entry + "\n")
+
+
+def get_entry_size(base_dir, dat_entry):
+    """Obtain the file size of a dat entry"""
+    dat_file = f"{dat_entry}_2.dat"
+    file_path = os.path.join(base_dir, dat_file)
+    return os.path.getsize(file_path)
+
+
+def with_progress(elements):
+    """A quick and easy generator for displaying progress while iterating"""
+    total_elems = len(elements)
+    report_every = math.ceil(total_elems / PROGRESS_STEPS)
+    logging.info(f"Processing {total_elems} entries")
+    for i, elem in enumerate(elements):
+        if i % report_every == 0:
+            print(f"{i:10} [{i*100/total_elems:3.0f}%]", file=sys.stderr)
+        yield elem
+
+
+def show_histogram(buckets, n_bins=DEFAULT_HISTOGRAM_NBINS):
+    """A simple histogram CLI visualizer"""
+    logging.info("Histogram of the Machine accumulated data")
+    freq, bins = numpy.histogram(buckets, bins=n_bins)
+    bin_start = bins[0]
+    for count, bin_end in zip(freq, bins[1:]):
+        if count:
+            print(f"  [{bin_start/(1024*1024):5.0f} - {bin_end/(1024*1024):5.0f}]: {count:0d}")
+        bin_start = bin_end
```

Comment:
These util functions are defined and the same in each script, maybe save them in one place and import for the other?